src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -7291,52 +7379,58 @@ and TcAssertExpr cenv overallTy env (m: range) tpenv x =
-              let withExpr, tpenv = TcExpr cenv (MustEqual overallTy) env tpenv origExpr
-              Some withExpr, tpenv
+    let withExprOptChecked, tpenv =
+        match withExprOpt with
+        | None -> None, tpenv
+        | Some (origExpr, _) ->
+            match inherits with
+            | Some (_, _, mInherits, _, _) -> error(Error(FSComp.SR.tcInvalidRecordConstruction(), mInherits))
+            | None ->
+                let withExpr, tpenv = TcExpr cenv (MustEqual overallTy) env tpenv origExpr
+                Some withExpr, tpenv
 
-    let hasOrigExpr = withExprOpt.IsSome
+    let hasOrigExpr = withExprOptChecked.IsSome
 
-    let fldsList =
+    let fldsList, containsNestedUpdates =
         let flds =
-            [
+            synRecdFields
+            |> List.choose (fun (SynExprRecordField (fieldName = (synLongId, isOk); expr = v)) ->
                 // if we met at least one field that is not syntactically correct - raise ReportedError to transfer control to the recovery routine
-                for SynExprRecordField(fieldName=(synLongId, isOk); expr=v) in synRecdFields do
-                    if not isOk then
-                        // raising ReportedError None transfers control to the closest errorRecovery point but do not make any records into log
-                        // we assume that parse errors were already reported
-                        raise (ReportedError None)
+                if not isOk then
+                    // raising ReportedError None transfers control to the closest errorRecovery point but do not make any records into log
+                    // we assume that parse errors were already reported
+                    raise (ReportedError None)
 
-                    yield (List.frontAndBack synLongId.LongIdent, v)
-            ]
+                match withExprOpt with
+                | Some withExpr ->
+                    match synLongId.LongIdent with
+                    | [] -> None
+                    | [ id ] -> Some (false, ([], id), v)
+                    | lid -> TransformAstForNestedUpdates cenv env overallTy lid v withExpr
+                | _ -> Some (false, List.frontAndBack synLongId.LongIdent, v))
+
+        let flds = if hasOrigExpr then GroupUpdatesToNestedFields flds else flds
 
         match flds with
-        | [] -> []
+        | [] -> [], false
         | _ ->
-            let tinst, tcref, _, fldsList = BuildFieldMap cenv env hasOrigExpr overallTy flds mWholeExpr
+            let tinst, tcref, _, fldsList = BuildFieldMap cenv env hasOrigExpr overallTy (flds |> List.map (fun (_, x, y) -> x, y)) mWholeExpr
             let gtyp = mkAppTy tcref tinst
             UnifyTypes cenv env mWholeExpr overallTy gtyp
 
             [ for n, v in fldsList do
                 match v with
                 | Some v -> yield n, v
-                | None -> () ]
+                | None -> () ], flds |> List.exists (fun (isNestedUpdate, _, _) -> isNestedUpdate)
```

Comment:
I am considering setting a mutable variable on L7415 to avoid these 2 `flds` iterations to be honest....